File: packages/scheduling-engine/src/analysis/transparency-report.ts
// ═══════════════════════════════════════════════════════════
//  INCOMPOL PLAN — Transparency Report Builder
//  Produces auditable per-order justifications for scheduling.
//
//  For feasible ops: WHY production starts when it does.
//  For infeasible ops: WHY it can't meet the deadline + suggestions.
//
//  Pure function -- no React, no side effects.
// ═══════════════════════════════════════════════════════════

import { DAY_CAP } from '../constants.js';
import type { Block, ZoneShiftDemand } from '../types/blocks.js';
import type { DecisionEntry } from '../types/decisions.js';
import type { EOp, ETool } from '../types/engine.js';
import type { InfeasibilityEntry } from '../types/infeasibility.js';
import type { CapacityLogEntry, DeficitEvolution, WorkContent } from '../types/scoring.js';
import type { OperationDeadline } from '../types/shipping.js';
import type {
  FailureJustification,
  OrderJustification,
  StartReason,
  TransparencyReport,
} from '../types/transparency.js';
import type { TwinValidationReport } from '../types/twin.js';
import type { WorkforceForecast } from '../types/workforce.js';
import { getBlockProductionForOp, getBlocksForOp } from '../utils/block-production.js';

// ── Helpers ──────────────────────────────────────────────────

/**
 * Determine the start reason for a feasible operation
 * based on scoring metrics and block data.
 */
function determineStartReason(
  wc: WorkContent,
  de: DeficitEvolution,
  deadline: OperationDeadline | undefined,
  blocks: Block[],
): StartReason {
  // If deficit already exists at day 0 (backlog)
  if (de.firstDeficitDay === 0 || de.dailyDeficit[0] < 0) {
    return 'deficit_elimination';
  }

  // If slack is critical (< 1 day)
  if (deadline) {
    const slackMin = deadline.latestFinishAbs - wc.workContentMin;
    if (slackMin < DAY_CAP) {
      return 'urgency_slack_critical';
    }
  }

  // If density is high (needs > 50% of available capacity in remaining time)
  if (deadline && wc.daysRequired > 0) {
    const availableDays = deadline.shippingDayIdx + 1;
    const density = wc.daysRequired / availableDays;
    if (density > 0.5) {
      return 'density_heavy_load';
    }
  }

  // If the block was moved by load leveling
  const opBlocks = blocks.filter((b) => b.opId === wc.opId && b.type === 'ok');
  if (opBlocks.some((b) => b.isLeveled)) {
    return 'future_load_relief';
  }

  // Default: free window
  return 'free_window_available';
}

/**
 * Collect shifts used per day from blocks for an operation.
 */
function collectShiftsPerDay(blocks: Block[], opId: string, nDays: number): ('X' | 'Y' | 'Z')[][] {
  const perDay: ('X' | 'Y' | 'Z')[][] = [];
  for (let d = 0; d < nDays; d++) {
    const shifts = new Set<'X' | 'Y' | 'Z'>();
    for (const b of blocks) {
      if (b.opId === opId && b.dayIdx === d && b.type === 'ok') {
        shifts.add(b.shift);
      }
    }
    perDay.push([...shifts].sort());
  }
  return perDay;
}

// ── Main export ──────────────────────────────────────────────

/**
 * Build a transparency report with per-order justifications.
 *
 * @param blocks           - Final scheduled blocks
 * @param ops              - All operations
 * @param toolMap          - Tool lookup by ID
 * @param deadlines        - Shipping deadlines per op
 * @param workContents     - Work content per op
 * @param deficits         - Deficit evolution per op
 * @param infeasibilities  - Infeasibility entries
 * @param decisions        - All decision entries (for capacity log extraction)
 * @param twinValidation   - Twin validation report (optional)
 * @returns TransparencyReport
 */
export function buildTransparencyReport(
  blocks: Block[],
  ops: EOp[],
  toolMap: Record<string, ETool>,
  deadlines: Map<string, OperationDeadline>,
  workContents: Map<string, WorkContent>,
  deficits: Map<string, DeficitEvolution>,
  infeasibilities: InfeasibilityEntry[],
  decisions: DecisionEntry[],
  twinValidation?: TwinValidationReport,
  workforceWarnings?: ZoneShiftDemand[],
  workforceForecast?: WorkforceForecast,
): TransparencyReport {
  const orderJustifications: OrderJustification[] = [];
  const failureJustifications: FailureJustification[] = [];

  // Extract capacity log from decisions
  const capacityLog: CapacityLogEntry[] = decisions
    .filter((d) => d.type === 'CAPACITY_COMPUTATION')
    .map((d) => d.metadata as unknown as CapacityLogEntry);

  // Build infeasible op set
  const infeasibleOpIds = new Set(infeasibilities.map((e) => e.opId));

  // Compute nDays from max dailyDeficit length
  const nDays = Math.max(...Array.from(deficits.values()).map((d) => d.dailyDeficit.length), 0);

  for (const op of ops) {
    const wc = workContents.get(op.id);
    const de = deficits.get(op.id);
    const deadline = deadlines.get(op.id);
    const tool = toolMap[op.t];

    // Skip ops with no demand
    const totalDemand = op.atr + op.d.reduce((s, v) => s + Math.max(v, 0), 0);
    if (totalDemand <= 0) continue;

    if (infeasibleOpIds.has(op.id)) {
      // ── Failure justification ──
      const entry = infeasibilities.find((e) => e.opId === op.id)!;
      const pH = tool?.pH ?? 0;
      const oee = tool?.oee ?? 0.66;
      const effectivePH = pH * oee;

      // Missing capacity — twin-aware production attribution
      const produced = getBlockProductionForOp(blocks, op.id);
      const missingPcs = totalDemand - produced;
      const missingHours = effectivePH > 0 ? missingPcs / effectivePH : 0;

      // First impossible moment: deadline abs or latest block end
      let firstImpossibleMoment = 0;
      if (deadline) {
        firstImpossibleMoment = deadline.latestFinishAbs;
      }

      // Suggestions
      const suggestions: string[] = [];
      if (entry.suggestion) {
        suggestions.push(...entry.suggestion.split('; '));
      }

      failureJustifications.push({
        opId: op.id,
        constraintsViolated: [entry.reason],
        firstImpossibleMoment,
        missingCapacityHours: Math.round(missingHours * 100) / 100,
        missingCapacityPieces: missingPcs,
        suggestions,
      });
    } else if (wc && de) {
      // ── Order justification ──
      const okBlocks = blocks.filter((b) => b.opId === op.id && b.type === 'ok');
      // Twin-aware production attribution
      const totalProduced = getBlockProductionForOp(blocks, op.id);

      // Allocated hours per day
      const totalProdMin = okBlocks.reduce((s, b) => s + b.prodMin, 0);
      const daysWithProd = new Set(okBlocks.map((b) => b.dayIdx)).size;
      const allocatedHoursPerDay = daysWithProd > 0 ? totalProdMin / daysWithProd / 60 : 0;

      // Shifts per day
      const shiftsUsedPerDay = collectShiftsPerDay(blocks, op.id, nDays);

      // Capacity
      const pH = wc.pH;
      const oee = wc.oee;
      const availableHoursPerDay = DAY_CAP / 60;
      const capacityPcsPerDay = pH * oee * availableHoursPerDay;

      // Start reason
      const startReason = determineStartReason(wc, de, deadline, blocks);

      // Initial deficit: if deficit exists at day 0, it's the negative of dailyDeficit[0]
      const initialDeficit =
        de.dailyDeficit.length > 0 && de.dailyDeficit[0] < 0 ? -de.dailyDeficit[0] : 0;

      // Twin co-production metadata
      const twinBlks = getBlocksForOp(blocks, op.id).filter((b) => b.isTwinProduction);
      const isTwinProd = twinBlks.length > 0;
      let twinPartnerSku: string | undefined;
      let twinOutputs: Array<{ sku: string; qty: number }> | undefined;
      if (isTwinProd) {
        // Aggregate outputs per SKU across all twin blocks
        const qtyBySku: Record<string, number> = {};
        for (const tb of twinBlks) {
          if (tb.outputs) {
            for (const o of tb.outputs) {
              qtyBySku[o.sku] = (qtyBySku[o.sku] ?? 0) + o.qty;
            }
          }
        }
        twinOutputs = Object.entries(qtyBySku).map(([sku, qty]) => ({ sku, qty }));
        // Partner = the other SKU in outputs
        const partner = twinBlks[0]?.outputs?.find((o) => o.opId !== op.id);
        twinPartnerSku = partner?.sku;
      }

      orderJustifications.push({
        opId: op.id,
        initialStock: de.initialStock,
        initialDeficit,
        deficitEvolution: [...de.dailyDeficit],
        pH,
        oee,
        capacityPcsPerDay: Math.round(capacityPcsPerDay * 100) / 100,
        allocatedHoursPerDay: Math.round(allocatedHoursPerDay * 100) / 100,
        shiftsUsedPerDay,
        startReason,
        feasible: true,
        totalProduced,
        totalDemand,
        ...(isTwinProd ? { isTwinProduction: true, twinPartnerSku, twinOutputs } : {}),
      });
    }
  }

  return {
    orderJustifications,
    failureJustifications,
    capacityLog,
    twinValidationReport: twinValidation,
    workforceWarnings,
    workforceForecast,
  };
}

File: apps/backend/src/domain/scheduling/analysis/transparency_report.py
"""Transparency report — port of analysis/transparency-report.ts.

Per-order justifications explaining WHY operations are scheduled when/where.
"""

from __future__ import annotations

from dataclasses import dataclass, field

from ..constants import DAY_CAP, DEFAULT_OEE
from ..types import Block, DecisionEntry, EOp, ETool, InfeasibilityEntry


@dataclass
class OrderJustification:
    op_id: str
    sku: str = ""
    initial_stock: int = 0
    initial_deficit: int = 0
    pH: int = 0
    oee: float = DEFAULT_OEE
    capacity_pcs_per_day: float = 0
    allocated_hours_per_day: float = 0
    start_reason: str = "free_window_available"
    feasible: bool = True
    total_produced: int = 0
    total_demand: int = 0
    is_twin_production: bool = False
    twin_partner_sku: str | None = None


@dataclass
class FailureJustification:
    op_id: str
    sku: str = ""
    constraints_violated: list[str] = field(default_factory=list)
    first_impossible_moment: int = 0
    missing_capacity_hours: float = 0
    missing_capacity_pieces: int = 0
    suggestions: list[str] = field(default_factory=list)


@dataclass
class CapacityLogEntry:
    op_id: str
    tool_id: str = ""
    machine_id: str = ""
    oee_value: float = DEFAULT_OEE
    oee_source: str = "default"
    pieces_per_hour: int = 0
    available_hours_per_day: float = 0
    resulting_capacity_pcs_per_day: float = 0
    work_content_hours: float = 0
    days_required: float = 0


@dataclass
class TransparencyReport:
    order_justifications: list[OrderJustification] = field(default_factory=list)
    failure_justifications: list[FailureJustification] = field(default_factory=list)
    capacity_log: list[CapacityLogEntry] = field(default_factory=list)


def _get_block_production_for_op(blocks: list[Block], op_id: str) -> int:
    total = 0
    for b in blocks:
        if b.type != "ok" or b.qty <= 0:
            continue
        if b.op_id == op_id:
            total += b.qty
        if b.outputs:
            for out in b.outputs:
                if out.op_id == op_id:
                    total += out.qty
    return total


def _determine_start_reason(
    blocks: list[Block],
    op: EOp,
    demand: int,
) -> str:
    """Determine why production started when it did."""
    # Deficit at start (backlog)
    if op.atr > 0:
        return "deficit_elimination"

    # Check if any block is leveled
    for b in blocks:
        if b.op_id == op.id and b.type == "ok" and b.is_leveled:
            return "future_load_relief"

    return "free_window_available"


def build_transparency_report(
    blocks: list[Block],
    ops: list[EOp],
    tool_map: dict[str, ETool],
    infeasibilities: list[InfeasibilityEntry] | None = None,
    decisions: list[DecisionEntry] | None = None,
) -> TransparencyReport:
    """Build transparency report with per-order justifications."""
    infeas = infeasibilities or []
    infeas_op_ids = set(e.op_id for e in infeas)

    order_justifications: list[OrderJustification] = []
    failure_justifications: list[FailureJustification] = []
    capacity_log: list[CapacityLogEntry] = []

    for op in ops:
        demand = sum(max(v, 0) for v in op.d) + max(op.atr, 0)
        if demand <= 0:
            continue

        tool = tool_map.get(op.t)
        pH = tool.pH if tool else 0
        oee = tool.oee if tool and tool.oee else DEFAULT_OEE

        if op.id in infeas_op_ids:
            # Failure justification
            entry = next((e for e in infeas if e.op_id == op.id), None)
            effective_pH = pH * oee if pH > 0 else 0
            produced = _get_block_production_for_op(blocks, op.id)
            missing_pcs = max(0, demand - produced)
            missing_hours = missing_pcs / effective_pH if effective_pH > 0 else 0

            suggestions = []
            if entry and entry.suggestion:
                suggestions = [s.strip() for s in entry.suggestion.split(";")]

            failure_justifications.append(
                FailureJustification(
                    op_id=op.id,
                    sku=op.sku,
                    constraints_violated=[entry.reason] if entry else [],
                    missing_capacity_hours=round(missing_hours, 2),
                    missing_capacity_pieces=missing_pcs,
                    suggestions=suggestions,
                )
            )
        else:
            # Order justification
            produced = _get_block_production_for_op(blocks, op.id)
            ok_blocks = [b for b in blocks if b.op_id == op.id and b.type == "ok"]
            total_prod_min = sum(b.prod_min for b in ok_blocks)
            days_with_prod = len(set(b.day_idx for b in ok_blocks))
            alloc_hours = (total_prod_min / days_with_prod / 60.0) if days_with_prod > 0 else 0
            cap_pcs_day = pH * oee * (DAY_CAP / 60.0) if pH > 0 else 0
            start_reason = _determine_start_reason(blocks, op, demand)

            is_twin = any(b.is_twin_production and b.type == "ok" for b in ok_blocks)

            order_justifications.append(
                OrderJustification(
                    op_id=op.id,
                    sku=op.sku,
                    initial_stock=op.stk if op.stk else 0,
                    initial_deficit=max(op.atr, 0),
                    pH=pH,
                    oee=oee,
                    capacity_pcs_per_day=round(cap_pcs_day, 1),
                    allocated_hours_per_day=round(alloc_hours, 2),
                    start_reason=start_reason,
                    total_produced=produced,
                    total_demand=demand,
                    is_twin_production=is_twin,
                )
            )

        # Capacity log entry
        if tool and pH > 0:
            oee_source = "tool" if tool.oee else "default"
            cap_pcs = pH * oee * (DAY_CAP / 60.0)
            work_hours = demand / (pH * oee) if pH * oee > 0 else 0
            days_req = work_hours / (DAY_CAP / 60.0) if DAY_CAP > 0 else 0

            capacity_log.append(
                CapacityLogEntry(
                    op_id=op.id,
                    tool_id=op.t,
                    machine_id=op.m,
                    oee_value=oee,
                    oee_source=oee_source,
                    pieces_per_hour=pH,
                    available_hours_per_day=DAY_CAP / 60.0,
                    resulting_capacity_pcs_per_day=round(cap_pcs, 1),
                    work_content_hours=round(work_hours, 2),
                    days_required=round(days_req, 2),
                )
            )

    return TransparencyReport(
        order_justifications=order_justifications,
        failure_justifications=failure_justifications,
        capacity_log=capacity_log,
    )
